Memory: Use a table based lookup scheme to read from memory regions

yuriks opened the pull request:
This replaces the tree of if-else in the memmap functions with a table of pointers which can be updated during runtime.

This is mostly ground work for proper kernel address space management, but in the meantime, gives a small ~3% speed up in general.

bunnei commented on src/core/hle/shared_page.h:
```diff
@@ -11,18 +11,45 @@
-void Read(T &var, const u32 addr);
-
-void Set3DSlider(float amount);
+// see http://3dbrew.org/wiki/Configuration_Memory#Shared_Memory_Page_For_ARM11_Processes
```
*See

bunnei commented on src/core/hle/shared_page.h:
```diff
@@ -11,18 +11,45 @@
+static_assert(sizeof(DateTime) == 0x20, "Datetime size is wrong");
+
+struct SharedPageDef {
+    // most of these names are taken from the 3dbrew page linked above.
```
*Most (sorry, comments that aren't capitalized kill me, I know, this isn't your doing)

bunnei commented on src/core/hle/shared_page.h:
```diff
@@ -11,18 +11,45 @@
+    // most of these names are taken from the 3dbrew page linked above.
+    u32_le   date_time_selector;         // 0
+    u8       running_hw;                 // 4
+    u8       mcu_hw_info;                // 5: don't know what the acronyms mean
```
This is a terrible line... "unknown" and get rid of comment?

yuriks replied:
I'm just gonna get rid of the comment, it's pretty obvious to me it means "microcontroller hardware info".

Subv replied:
Please add a comment saying that, the MCU = Microcontroller Hardware Info wasn't obvious at all to me :P

bunnei commented on src/core/hle/shared_page.h:
```diff
@@ -11,18 +11,45 @@
+    DateTime date_time_0;                // 20
+    DateTime date_time_1;                // 40
+    u8       wifi_macaddr[6];            // 60
+    u8       wifi_unknown1;              // 66: 3dbrew says these are "Likely wifi hardware related"
```
3DBrew

Subv commented on src/core/memory.cpp:
```diff
@@ -0,0 +1,187 @@
+// Copyright 2014 Citra Emulator Project
```
2015

Subv commented on src/core/memory.h:
```diff
@@ -0,0 +1,125 @@
+// Copyright 2014 Citra Emulator Project
```
2015, same in most other files

Subv commented on src/core/memory.h:
```diff
@@ -0,0 +1,125 @@
+
+namespace Memory {
+
+const u32 PAGE_SIZE = 0x1000;
```
Is this arbitrary or is it an official value?

yuriks replied:
It's the page size used by the ARM architecture... (of the sub-pages, anyway, real ARM uses a 2-level pagetable.)

archshift replied:
It's an official value.

Subv commented:
Overall LGTM after the commets are addressed

yuriks commented:
Updated.

@Subv: I didn't change all the copyrights since most of those files we're just moved from existing code, and I don't think you should update the date on those situations, so I just changed it on `memory.cpp` which I mostly re-wrote from scratch. Maybe @neobrain has an opinion on it.

bunnei commented on src/core/memory.cpp:
```diff
@@ -0,0 +1,187 @@
+enum class PageType {
+    Unmapped = 0,
+    Memory = 1,
+    Special = 2,
```
What does "Special" mean, except that it's used below for IO?

bunnei commented:
Looks good!

bunnei commented:
This is pending some documentation before being ready for merge

yuriks commented:
Updated. Someone just quickly proofread the last commit before I squash.

Subv commented:
LGTM
